packages/store/src/patterns/encodeKey.js

```diff
@@ -79,69 +79,101 @@ const dbEntryKeyToNumber = k => {
   }
   asBits[0] = bits;
   const result = asNumber[0];
+
+  // Normalize -0 to 0
+  // XXX We never expect to decode -0, should this instead assert that?
   if (is(result, -0)) {
     return 0;
   }
   return result;
 };
 
-// BigInts are encoded as keys as follows:
-//   `${prefix}${length}:${encodedNumber}`
-// Where:
-//
-//   ${prefix} is either 'n' or 'p' according to whether the BigInt is negative
-//      or positive ('n' is less than 'p', so negative BigInts will sort below
-//      positive ones)
-//
-//   ${encodedNumber} is the value of the BigInt itself, encoded as a decimal
-//      number.  Positive BigInts use their normal decimal representation (i.e.,
-//      what is returned when you call `toString()` on a BigInt).  Negative
-//      BigInts are encoded as the unpadded 10s complement of their value; in
-//      this encoding, all negative values that have same number of digits will
-//      sort lexically in the inverse order of their numeric value (which is to
-//      say, most negative to least negative).
-//
-//   ${length} is the decimal representation of the width (i.e., the count of
-//      digits) of the BigInt.  This length value is then zero padded to a fixed
-//      number (currently 10) of digits.  Note that the fixed width length field
-//      means that we cannot encode BigInts whose values have more than 10**10
-//      digits, but we are willing to live with this limitation since we could
-//      never store such large numbers anyway.  The length field is used in lieu
-//      of zero padding the BigInts themselves for sorting, which would be
-//      impractical for the same reason that storing large values directly would
-//      be.  The length is zero padded so that numbers are sorted within groups
-//      according to their decimal orders of magnitude in size and then these
-//      groups are sorted smallest to largest.
-//
-// This encoding allows all BigInts to be represented as ASCII strings that sort
-// lexicographically in the same order as the values of the BigInts themselves
-// would sort numerically.
-
-export const BIGINT_TAG_LEN = 10;
-const BIGINT_LEN_MODULUS = 10 ** BIGINT_TAG_LEN;
-
-const bigintToDBEntryKey = n => {
+// JavaScript bigints are encoded using a variant of Elias delta coding, with an
+// initial component for the length of the digit count as a unary string, a
+// second component for the decimal digit count, and a third component for the
+// decimal digits preceded by a gratuitous separating colon.
+// To ensure that the lexicographic sort order of encoded values matches the
+// numeric sort order of the corresponding numbers, the characters of the unary
+// prefix are different for negative values (type "n" followed by any number of
+// "#"s [which sort before decimal digits]) vs. positive and zero values (type
+// "p" followed by any number of "~"s [which sort after decimal digits]) and
```

Comment:
It would also be possible to use a single type character for all bigint values rather than "n" for negative and "p" for nonnegative, e.g.

n | encoded
-- | --
-1n | i#8:8
0n | i~1:0
1n | i~1:1
5n | i~1:5
42n | i~2:42
1234567890n | i~~10:1234567890

Reply:
There's no significant difference, it just seemed worth raising in case of perceived value from e.g. economy of type prefixes and/or one-to-one mappings. Assuming not, we'll stick with "n" and "p".

Reply:
good